### DISCUSSION TITLE
Confusion about TIMESTAMPTZ behavior: Detection of client timezone

### DISCUSSION THREAD

**Discussion by benrbowers** [supabase/supabase#22144] [Questions]
TL;DR: Supabase STORES and RETURNS timestamptz as UTC, which turns out is ideal as you can then just throw that into Date.getLocaleTimeString

I've seen most people online recommend `timestamptz` for keeping times consistent across clients. And many of those proponents list as one their reasons for doing so is that you can both insert in local time AND retrieve in local time. However, supabase seems to only have _half_ of that behavior.

When I insert a string of the form `YYYY-MM-DD hh:mm:ss` (notice no explicit timezone) into a column of the type `timestamptz`, the time is interpreted to have the timezone of the client connection and then converted to and stored in UTC, as expected. However, when I _query_ that column from the _same client_ I receive the timestamp in the stored UTC form. I then have to convert the time on the client _from_ UTC to local time. 

Is this the intended behavior? Supabase clearly has the capability to convert based on the client connection because **_it gets the initial conversion right,_** so this has me a little confused.

**Comment by benrbowers** [comment]
I've got it working as intended now. The trick was to EXPLICITLY state the timezone offset in the INSERT. I was mistaken about the initial conversion. It was simply taking 10 AM and storing it as 10 AM UTC. And then when I threw that into Date.toLocaleTimeString(), it gave me 6 AM which is what _would_ have been the proper UTC time, hence my confusion. 

I get the offset like this:
```
export function getTimezoneOffset() {
	const offsetMinutes = -new Date().getTimezoneOffset();

	let offsetHours: string | number = offsetMinutes / 60;
	const sign = offsetHours >= 0 ? '+' : '-';
	offsetHours = Math.abs(offsetHours);
	offsetHours = offsetHours < 10 ? '0' + offsetHours : offsetHours;

	let offsetRemainder: string | number = offsetMinutes % 60;
	offsetRemainder =
		offsetRemainder < 10 ? '0' + offsetRemainder : offsetRemainder;

	return `${sign}${offsetHours}:${offsetRemainder}`;
}

export function useTimezoneOffset() {
	return useMemo<string>(getTimezoneOffset, []);
}
```

Then I insert like this:
```
.insert({
    start_time: `${eventDate} ${startTime} ${tzOffset}`,
    end_time: `${endDate} ${endTime} ${tzOffset}`,
})
```

And then you do indeed get UTC on the retrieval, which is handled easily by `new Date(timestamp).getLocaleTimeString()`
